crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,206 @@
+        let object_store: Arc<dyn ObjectStore> = if destination.scheme() == "s3"
+            && destination.query().is_none()
+            && snapshots_options.additional_options.is_empty()
+        {
+            let aws_region = aws_config::load_defaults(BehaviorVersion::v2024_03_28())
+                .await
+                .region()
+                .context("Unable to determine AWS region to use with S3")?
+                .clone();
+
+            let store = AmazonS3Builder::new()
+                .with_url(destination.clone())
+                .with_region(aws_region.to_string())
+                .with_credentials(Arc::new(AwsSdkCredentialsProvider {
+                    credentials_provider: DefaultCredentialsChain::builder().build().await,
+                }))
+                .build()?;
+
+            Arc::new(store)
+        } else {
+            object_store::parse_url_opts(&destination, &snapshots_options.additional_options)?
+                .0
+                .into()
+        };
+
+        let prefix = destination.path().into();
+        Ok(SnapshotRepository {
+            object_store,
+            destination,
+            prefix,
+            staging_path: base_dir.clone().join("snapshot-staging"),
+        })
+    }
+
+    /// Write a partition snapshot to the snapshot repository.
+    pub(crate) async fn put(
+        &self,
+        partition_id: PartitionId,
+        metadata: &PartitionSnapshotMetadata,
+        snapshot_path: &Path,
+    ) -> anyhow::Result<()> {
+        let snapshot_id = metadata.snapshot_id;
+        let lsn = metadata.min_applied_lsn;
+
+        debug!(
+            %snapshot_id,
+            partition_id = ?partition_id,
+            %lsn,
+            "Publishing partition snapshot to: {}",
+            self.destination,
+        );
+
+        // All common object stores list objects in lexicographical order, with no option for
+        // reverse order. We inject an explicit sort key into the snapshot prefix to make sure that
+        // the latest snapshot is always first.
+        let inverted_sort_key = format!("{:016x}", u64::MAX - lsn.as_u64());
+
+        // The snapshot data / metadata key format is: [<base_prefix>/]<partition_id>/<sort_key>_<lsn>_<snapshot_id>.tar
+        let snapshot_key = match self.prefix.as_str() {
+            "" | "/" => format!(
+                "{partition_id}/{sk}_{lsn}_{snapshot_id}.tar",
+                sk = inverted_sort_key,
+                lsn = metadata.min_applied_lsn,
+            ),
+            prefix => format!(
+                "{trimmed_prefix}/{partition_id}/{sk}_{lsn}_{snapshot_id}.tar",
+                trimmed_prefix = prefix.trim_start_matches('/').trim_end_matches('/'),
+                sk = inverted_sort_key,
+            ),
+        };
+
+        let staging_path = self.staging_path.clone();
+        tokio::fs::create_dir_all(&staging_path).await?;
```

Comment:
Could this happen when we create the `SnapshotRepository` instead of on every `put`?